// @hsui/core/lib/fetch.js
import _objectSpread from "@babel/runtime/helpers/esm/objectSpread2";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import { ajax } from '@hsui/sdk';
import { extend } from './utils';
import { MiddlewareManager, beforeRequestSend, afterRequestSend } from './middleware';
import { log, logger } from './log';

function final_(error, response) {
  if (error) {
    return Promise.reject(error);
  }

  return Promise.resolve(response);
}

var defaultConfig = {
  method: 'post',
  data: {},
  timeout: 5000,
  headers: {},
  withCredentials: false
};

var Fetch = function Fetch() {
  var _config = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

  _classCallCheck(this, Fetch);

  extend(defaultConfig, _config);
  beforeRequestSend.final(function (config) {
    return new Promise(function (resolve, reject) {
      var error, response;
      ajax(config).then(function (res) {
        return response = res;
      }).catch(function (err) {
        return error = err;
      }).finally(function () {
        afterRequestSend.runAndReturn(error, response).then(resolve).catch(reject);
      });
    });
  });
  afterRequestSend.final(final_);

  function fetch(config) {
    log.debug('Fetch 实例初始化选项', _objectSpread(_objectSpread({}, defaultConfig), config));
    return beforeRequestSend.runAndReturn(_objectSpread(_objectSpread({}, defaultConfig), config));
  }

  fetch.get = function (url, config) {
    if (config) {
      extend(config, {
        url: url,
        method: 'get'
      });
    } else {
      config = {
        url: url,
        method: 'get'
      };
    }

    return fetch(config);
  };

  fetch.post = function (url, data, config) {
    if (config) {
      extend(config, {
        url: url,
        method: 'post',
        data: data
      });
    } else {
      config = {
        url: url,
        method: 'post',
        data: data
      };
    }

    return fetch(config);
  };

  return fetch;
};

export { Fetch as default };
export function initFetch() {
  var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
      _ref$defaultConfig = _ref.defaultConfig,
      defaultConfig = _ref$defaultConfig === void 0 ? {
    method: 'post',
    data: {},
    timeout: 5000,
    headers: {},
    withCredentials: false
  } : _ref$defaultConfig,
      _ref$beforeRequestSen = _ref.beforeRequestSendMiddlewares,
      beforeRequestSendMiddlewares = _ref$beforeRequestSen === void 0 ? [] : _ref$beforeRequestSen,
      _ref$afterRequestSend = _ref.afterRequestSendMiddlewares,
      afterRequestSendMiddlewares = _ref$afterRequestSend === void 0 ? [] : _ref$afterRequestSend;

  var beforeRequestSend = new MiddlewareManager('before-request-send');
  beforeRequestSend.concat(beforeRequestSendMiddlewares);
  var afterRequestSend = new MiddlewareManager('after-request-send');
  afterRequestSend.concat(afterRequestSendMiddlewares);
  afterRequestSend.final(final_);
  beforeRequestSend.final(function (config) {
    return new Promise(function (resolve, reject) {
      var error, response;
      ajax(config).then(function (res) {
        return response = res;
      }).catch(function (err) {
        return error = err;
      }).finally(function () {
        afterRequestSend.runAndReturn(error, response).then(resolve).catch(reject);
      });
    });
  });

  var fetch = function fetch(config) {
    log.debug('Fetch 实例初始化选项', _objectSpread(_objectSpread({}, defaultConfig), config));
    return beforeRequestSend.runAndReturn(_objectSpread(_objectSpread({}, defaultConfig), config));
  };

  fetch.get = function (url, config) {
    if (config) {
      extend(config, {
        url: url,
        method: 'get'
      });
    } else {
      config = {
        url: url,
        method: 'get'
      };
    }

    return fetch(config);
  };

  fetch.post = function (url, data, config) {
    if (config) {
      extend(config, {
        url: url,
        method: 'post',
        data: data
      });
    } else {
      config = {
        url: url,
        method: 'post',
        data: data
      };
    }

    return fetch(config);
  };

  return fetch;
}
